lib/IRGen/GenFunc.cpp

```diff
@@ -1244,7 +1244,20 @@ static llvm::Function *emitPartialApplicationForwarder(IRGenModule &IGM,
     // cast to the result type - it could be substituted.
     if (origConv.getSILResultType().hasTypeParameter()) {
       auto ResType = fwd->getReturnType();
-      callResult = subIGF.Builder.CreateBitCast(callResult, ResType);
+      if (auto *structType = dyn_cast<llvm::StructType>(ResType)) {
+        // Cast all struct elements to the desired type.
+        llvm::Value *castResult = llvm::UndefValue::get(structType);
+        for (auto i : range(structType->getStructNumElements())) {
+          auto desiredEltTy = structType->getElementType(i);
+          auto elt = subIGF.Builder.CreateExtractValue(callResult, {i});
+          auto castElt = subIGF.Builder.CreateBitCast(elt, desiredEltTy);
```

Comment:
Don't you need a recursion here?

Maybe emitCastToSubstSchema() can be used here?

Reply:
I don't know enough IRGen. Can any element be a nested struct? I assumed everything has been exploded at this stage.

Reply:
Everything is going to be exploded. When we form a direct return value we call getScalarResultType (among other things) which makes this assumption.

Reply:
You could use ``IRGenFunction::coerceValue`` instead.

```
    // If the result type is dependent on a type parameter we might have to
    // cast to the result type - it could be substituted.
    if (origConv.getSILResultType().hasTypeParameter()) {
      auto ResType = fwd->getReturnType();
      if (ResType != callResult->getType())
        callResult =
          subIGF.coerceValue(callResult, ResType, subIGF.IGM.DataLayout);
    }
    subIGF.Builder.CreateRet(callResult);
```

Reply:
coerceValue() looks much cleaner, thanks Arnold!

Reply:
Done. This is much nicer :)